File: Filters and Face Detection/Filters and Face Detection/Classes/PhotoFilter.m
//
//  PhotoFilter.m
//  Filters and Face Detection
//
//  Created by Saurabh Jain on 10/8/15.
//  Copyright © 2015 Saurabh Jain. All rights reserved.
//

#import "PhotoFilter.h"

@implementation PhotoFilter

#pragma mark - Helper Function

+ (void) applyFilter:(CIFilter *)filter toImage:(UIImage *)image withBlock:(void (^)(UIImage * image))block
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        // Create CGImage
        CIImage *inputImage  = [CIImage imageWithCGImage:[image CGImage]];
        
        // Create a context
        CIContext *context = [CIContext contextWithOptions:nil];
        
        // Set the input image to the filter
        [filter setValue:inputImage forKey:kCIInputImageKey];
        
        // Get the resultant image
        CIImage *resultImage = [filter valueForKey:kCIOutputImageKey];
        
        // CGImage from CIImage
        CGImageRef filteredCGImage = [context createCGImage:resultImage fromRect:[resultImage extent]];
        
        // UIImage
        UIImage *filteredImage =  [UIImage imageWithCGImage:filteredCGImage];
        
        if (block) {
            block(filteredImage);
        }
    });
}

@end

File: Phantom/Phantom/Classes/Helper/PhotoFilter.swift
//
//  PhotoFilter.swift
//  Phantom
//
//  Created by Saurabh Jain on 11/18/15.
//  Copyright © 2015 Saurabh Jain. All rights reserved.
//

import UIKit

protocol PhotoFilterProtocol: class {
    func applyFilter(filter: CIFilter, onImage image: UIImage, withCompletionBlock block:(UIImage?) -> ())
}

class PhotoFilter: PhotoFilterProtocol {

    // Context object
    private var context = CIContext()
    
    // Singleton
    private init() {}
    static let sharedFilter = PhotoFilter()
    
    // MARK: PhotoFilter Protocol
    
    func applyFilter(filter: CIFilter, onImage image: UIImage, withCompletionBlock block:(UIImage?) -> ()) {
        
        // get the CGimage
        guard let cgimage = image.CGImage else  {
            block(nil)
            return
        }
        
        // Do the filtering on a background thread
        dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INTERACTIVE, 0)) {
            // Get the CIImage
            let ciimage = CIImage(CGImage: cgimage)
            
            // Input to filter
            filter.setValue(ciimage, forKey: kCIInputImageKey)
            
            let orientation = image.imageOrientation
            let scale = image.scale
            
            // Get the output image
            if let img = filter.outputImage {
                let rect = img.extent
                let image = self.context.createCGImage(img, fromRect: rect)
                block(UIImage(CGImage: image, scale: scale, orientation: orientation))
            } else {
                block(nil)
            }
        }
        
    }
    
}
